controller/src/main/java/io/pravega/controller/server/bucket/BucketManagerLeader.java

```diff
@@ -0,0 +1,210 @@
+                case ERROR:
+                    // This event should be due to ZK connection errors and would have been received by the monitor too,
+                    // hence not handling it explicitly here.
+                    log.error("{}: Received error event when monitoring the pravega host cluster, ignoring...", serviceType);
+                    break;
+            }
+        });
+
+        // Keep looping here as long as possible to stay as the leader and exclusively monitor the pravega cluster.
+        while (true) {
+            try {
+                if (suspended.get()) {
+                    log.info("{}: Monitor is suspended, waiting for notification to resume.", serviceType);
+                    suspendMonitor.await();
+                    log.info("{}: Resuming monitor.", serviceType);
+                }
+
+                controllerChange.await();
+                log.info("{}: Received distribute buckets.", serviceType);
+
+                // Wait here until distribution can be performed.
+                waitForReDistribute();
+
+                // Clear all events that has been received until this point since this will be included in the current
+                // distribution operation.
+                controllerChange.reset();
+                triggerDistribution();
+            } catch (InterruptedException e) {
+                log.warn("{}: Leadership interrupted, releasing monitor thread.", serviceType);
+                Thread.currentThread().interrupt();
+                // Stop watching the pravega cluster.
+                synchronized (lock) {
+                    pravegaServiceCluster.close();
+                }
+                throw e;
+            } catch (Exception e) {
+                // We will not release leadership if in suspended mode.
+                if (!suspended.get()) {
+                    log.warn("{}: Failed to perform distribution, relinquishing leadership.", serviceType);
+
+                    // Stop watching the pravega cluster.
+                    synchronized (lock) {
+                        pravegaServiceCluster.close();
+                    }
+                    throw e;
+                }
+            }
+        }
+
+    }
+
+    /**
+     * Blocks until the redistribute interval. This wait serves multiple purposes:
+     * -- Ensure redistribution does not happen in quick succession since its a costly cluster level operation.
+     * -- Clubs multiple host events into one to reduce redistribution operations. For example:
+     *      Fresh cluster start, cluster/multi-host/host restarts, etc.
+     */
+    private void waitForReDistribute() throws InterruptedException, ExecutionException {
+        log.info("{}: Waiting for {} seconds before attempting to distribute.", serviceType,
+                minBucketRedistributionIntervalInSeconds.getSeconds());
+        Futures.delayedFuture(minBucketRedistributionIntervalInSeconds, executorService).get();
```

Comment:
This logic seems strange. It waits for an event such as a new host added, THEN it sleeps, then it redistributes. Shouldn't the idea be that this is conditional, it insures a minimum delay since startup or the last redistribution time, rather than just always waiting for a delay of X before reacting?

Reply:
As per our implementation, redistribution of buckets happens if and only if there is any change(Either new controller instance is up or existing controller instance fails) in pravega cluster. The idea here to wait for delay after getting the event is because of -

- Ensure redistribution does not happen in quick succession
- Clubs multiple host events into one to reduce redistribution operations.

Reply:
This is still wrong. What the code does now is:
* Wait for an event that should cause redistribution
* sleep for minBucketRedistributionIntervalInSeconds
* redistribute

What it should do is:
* Wait for an event that should cause redistribution
* IF minBucketRedistributionIntervalInSeconds has not passed since the last redistribution, wait until that time.
* redistribute